module/state_synchronization/indexer/indexer_test.go

```diff
@@ -0,0 +1,205 @@
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	mocks "github.com/stretchr/testify/mock"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data/cache"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data/mock"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	mempool "github.com/onflow/flow-go/module/mempool/mock"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+const testTimeout = 300 * time.Millisecond
+
+type indexerTest struct {
+	blocks        []*flow.Block
+	progress      *mockProgress
+	indexTest     *indexCoreTest
+	worker        *Indexer
+	executionData *mempool.ExecutionData
+	t             *testing.T
+}
+
+// newIndexerTest set up a jobqueue integration test with the worker.
+// It will create blocks fixtures with the length provided as availableBlocks, and it will set heights already
+// indexed to lastIndexedIndex value. Using run it should index all the remaining blocks up to all available blocks.
+func newIndexerTest(t *testing.T, availableBlocks int, lastIndexedIndex int) *indexerTest {
+	blocks := blocksFixture(availableBlocks)
+	// we use 5th index as the latest indexed height, so we leave 5 more blocks to be indexed by the indexer in this test
+	lastIndexedHeight := blocks[lastIndexedIndex].Header.Height
+	progress := &mockProgress{index: lastIndexedHeight}
+
+	indexerCoreTest := newIndexCoreTest(t, blocks, nil).
+		setLastHeight(func(t *testing.T) uint64 {
+			return progress.index
+		}).
+		useDefaultBlockByHeight().
+		initIndexer()
+
+	executionData := mempool.NewExecutionData(t)
+	exeCache := cache.NewExecutionDataCache(
+		mock.NewExecutionDataStore(t),
+		indexerCoreTest.indexer.headers,
+		nil,
+		nil,
+		executionData,
+	)
+
+	test := &indexerTest{
+		t:             t,
+		blocks:        blocks,
+		progress:      progress,
+		indexTest:     indexerCoreTest,
+		executionData: executionData,
+	}
+
+	test.worker = NewIndexer(
+		unittest.Logger(),
+		test.first().Header.Height,
+		testTimeout,
+		indexerCoreTest.indexer,
+		exeCache,
+		test.latestHeight,
+		progress,
+	)
+
+	return test
+}
+
+func (w *indexerTest) setBlockDataByID(f func(ID flow.Identifier) (*execution_data.BlockExecutionDataEntity, bool)) {
+	w.executionData.
+		On("ByID", mocks.AnythingOfType("flow.Identifier")).
+		Return(f)
+}
+
+func (w *indexerTest) latestHeight() (uint64, error) {
+	return w.last().Header.Height, nil
+}
+
+func (w *indexerTest) last() *flow.Block {
+	return w.blocks[len(w.blocks)-1]
+}
+
+func (w *indexerTest) first() *flow.Block {
+	return w.blocks[0]
+}
+
+func (w *indexerTest) run(ctx irrecoverable.SignalerContext, cancel context.CancelFunc) {
+	w.worker.Start(ctx)
+
+	unittest.RequireComponentsReadyBefore(w.t, testTimeout, w.worker.component)
+
+	w.worker.OnExecutionData(nil)
+
+	// give it a bit of time to process all the blocks
+	time.Sleep(testTimeout - 50)
```

Comment:
Are we able to replace this with a callback to be notified when all blocks all blocks are processed, so that we can stop waiting earlier. and also make the tests more reliable.

Reply:
I can look into this.

Reply:
Done